src/main/java/com/google/api/codegen/discovery/ViewModelProvider.java

```diff
@@ -0,0 +1,172 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.discovery;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.TreeMap;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.api.codegen.ApiaryConfig;
+import com.google.api.codegen.discovery.config.ApiaryConfigToSampleConfigConverter;
+import com.google.api.codegen.discovery.config.SampleConfig;
+import com.google.api.codegen.discovery.config.TypeNameGenerator;
+import com.google.api.codegen.discovery.transformer.MethodToViewTransformer;
+import com.google.api.codegen.rendering.CommonSnippetSetRunner;
+import com.google.api.codegen.viewmodel.ViewModel;
+import com.google.api.tools.framework.snippet.Doc;
+import com.google.protobuf.Method;
+
+/*
+ * Calls the MethodToViewTransformer on a method with the provided ApiaryConfig.
+ */
+public class ViewModelProvider implements DiscoveryProvider {
+
+  private final ApiaryConfig apiaryConfig;
+  private final CommonSnippetSetRunner snippetSetRunner;
+  private final MethodToViewTransformer methodToViewTransformer;
+  private final JsonNode sampleConfigOverrides;
+  private final TypeNameGenerator typeNameGenerator;
+  private final String outputRoot;
+
+  private ViewModelProvider(
+      ApiaryConfig apiaryConfig,
+      CommonSnippetSetRunner snippetSetRunner,
+      MethodToViewTransformer methodToViewTransformer,
+      JsonNode sampleConfigOverrides,
+      TypeNameGenerator typeNameGenerator,
+      String outputRoot) {
+    this.apiaryConfig = apiaryConfig;
+    this.snippetSetRunner = snippetSetRunner;
+    this.methodToViewTransformer = methodToViewTransformer;
+    this.sampleConfigOverrides = sampleConfigOverrides;
+    this.typeNameGenerator = typeNameGenerator;
+    this.outputRoot = outputRoot;
+  }
+
+  @Override
+  public Map<String, Doc> generate(Method method) {
+    // Before the transformer step, we generate the SampleConfig and apply overrides if available.
+    SampleConfig sampleConfig =
+        new ApiaryConfigToSampleConfigConverter(
+                Collections.singletonList(method), apiaryConfig, typeNameGenerator)
+            .convert();
+    if (sampleConfigOverrides != null) {
+      ObjectMapper mapper = new ObjectMapper();
+      JsonNode tree = mapper.valueToTree(sampleConfig);
+      merge((ObjectNode) tree, (ObjectNode) sampleConfigOverrides);
+      try {
+        sampleConfig = mapper.treeToValue(tree, SampleConfig.class);
+      } catch (Exception e) {
+        throw new RuntimeException("failed to parse config to node, this should never happen");
+      }
+    }
```

Comment:
Maybe to achieve a "balance of complexity" in this method, move the code to create the SampleConfig into a helper method; that way, this method just calls out to three things: 1. generate SampleConfig, 2. transform, 3. snippetSetRunner.generate